### ISSUE TITLE
MSBuild integration doesn't work for wxWidgets package

### ISSUE THREAD

**Issue by vadz** [Microsoft/vcpkg#3180]
### Synopsis

wxWidgets package installs one of its headers in `$VcpkgRoot\lib\mswu\wx\setup.h` where it's not found because the default headers path only contains `$VcpkgRoot\include`.

### How to reproduce

Perform the following steps (tested under Windows 7 x64 in 32 bit command line prompt, but shouldn't be OS-dependent):

1. Install vcpkg, by cloning it from the repository and running the bootstrap script.
1. Run `.\vcpkg.exe integrate install`.
1. Run `.\vcpkg.exe install wxwidgets`.
1. Create a new empty Win32 desktop project in MSVS 2017.
1. Create a new C++ source file containing `#include <wx/wx.h>` line.
1. Try to compile it, resulting in an error message similar to the following: `...\vcpkg\installed\x86-windows\include\wx\platform.h(148): fatal error C1083: Cannot open include file: 'wx/setup.h': No such file or directory`

### Solution

I'm not sure what would be the best way to solve this. Adding `$VcpkgRoot\lib\mswu` would fix this for the release builds, but in Debug the header is under `$VcpkgRoot\debug\lib\mswud`, so the include paths would need to differ depending on the configuration.

Alternatively, the file could be installed as `$VcpkgRoot\include\wx\setup.h` instead of its usual location because it doesn't actually depend on the configuration (in principle, it could, but with vcpkg build only `wxINSTALL_PREFIX` differs and this is not really important). This is probably the simplest thing to do.

**Comment by p1v0t** [comment]
> Adding $VcpkgRoot\lib\mswu …

@vadz What you exactly mean with it?

**Comment by vadz** [comment]
@p1v0t I mean to add this directory to the list of the include paths, i.e. the "Additional Include Directories" option.

**Comment by p1v0t** [comment]
Ohh, got it now, thanks ☺😃

**Comment by SeanFarrow** [comment]
I wonder whether there is a way to include subdirectories in the .targets file?

**Comment by degski** [comment]
@SeanFarrow 

The below is the complete .targets-file for the VS2017 integration (.\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets), that's where the include dirs get set:

    <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'Win32||'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x86-windows</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'Win32|Windows Store|10.0'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x86-uwp</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == '||'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x86-windows</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == '|Windows Store|10.0'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x86-uwp</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'x64||'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x64-windows</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'arm||'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm-windows</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'arm64||'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm64-windows</VcpkgTriplet>
      </PropertyGroup>  

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'x64|Windows Store|10.0'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x64-uwp</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'arm|Windows Store|10.0'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm-uwp</VcpkgTriplet>
      </PropertyGroup>

      <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'arm64|Windows Store|10.0'">
        <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
        <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm64-uwp</VcpkgTriplet>
      </PropertyGroup>  

      <PropertyGroup Condition="'$(VcpkgEnabled)' == 'true'">
        <VcpkgConfiguration Condition="'$(VcpkgConfiguration)' == ''">$(Configuration)</VcpkgConfiguration>
        <VcpkgNormalizedConfiguration Condition="$(VcpkgConfiguration.StartsWith('Debug'))">Debug</VcpkgNormalizedConfiguration>
        <VcpkgNormalizedConfiguration Condition="$(VcpkgConfiguration.StartsWith('Release')) or '$(VcpkgConfiguration)' == 'RelWithDebInfo' or '$(VcpkgConfiguration)' == 'MinSizeRel'">Release</VcpkgNormalizedConfiguration>
        <VcpkgRoot Condition="'$(VcpkgRoot)' == ''">$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), .vcpkg-root))\installed\$(VcpkgTriplet)\</VcpkgRoot>
      </PropertyGroup>

      <ItemDefinitionGroup Condition="'$(VcpkgEnabled)' == 'true'">
        <Link>
          <AdditionalDependencies Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug' and '$(VcpkgAutoLink)' != 'false'">%(AdditionalDependencies);$(VcpkgRoot)debug\lib\*.lib</AdditionalDependencies>
          <AdditionalDependencies Condition="'$(VcpkgNormalizedConfiguration)' == 'Release' and '$(VcpkgAutoLink)' != 'false'">%(AdditionalDependencies);$(VcpkgRoot)lib\*.lib</AdditionalDependencies>
          <AdditionalLibraryDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Release'">%(AdditionalLibraryDirectories);$(VcpkgRoot)lib;$(VcpkgRoot)lib\manual-link</AdditionalLibraryDirectories>
          <AdditionalLibraryDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug'">%(AdditionalLibraryDirectories);$(VcpkgRoot)debug\lib;$(VcpkgRoot)debug\lib\manual-link</AdditionalLibraryDirectories>
        </Link>
        <ClCompile>
          <AdditionalIncludeDirectories>%(AdditionalIncludeDirectories);$(VcpkgRoot)include</AdditionalIncludeDirectories>
        </ClCompile>
        <ResourceCompile>
          <AdditionalIncludeDirectories>%(AdditionalIncludeDirectories);$(VcpkgRoot)include</AdditionalIncludeDirectories>
        </ResourceCompile>
      </ItemDefinitionGroup>

      <Target Name="VcpkgTripletSelection" BeforeTargets="ClCompile">
        <Message Text="Using triplet &quot;$(VcpkgTriplet)&quot; from &quot;$(VcpkgRoot)&quot;" Importance="Normal" Condition="'$(VcpkgEnabled)' == 'true'"/>
        <Message Text="Not using Vcpkg because VcpkgEnabled is &quot;$(VcpkgEnabled)&quot;" Importance="Normal" Condition="'$(VcpkgEnabled)' != 'true'"/>
        <Message Text="Vcpkg is unable to link because we cannot decide between Release and Debug libraries. Please define the property VcpkgConfiguration to be 'Release' or 'Debug' (currently '$(VcpkgConfiguration)')." Importance="High" Condition="'$(VcpkgEnabled)' == 'true' and '$(VcpkgNormalizedConfiguration)' == ''"/>
      </Target>

      <Target Name="AppLocalFromInstalled" AfterTargets="CopyFilesToOutputDirectory" BeforeTargets="CopyLocalFilesOutputGroup;RegisterOutput" Condition="'$(VcpkgEnabled)' == 'true'">
        <WriteLinesToFile
        File="$(TLogLocation)$(ProjectName).write.1u.tlog"
        Lines="^$(TargetPath);$([System.IO.Path]::Combine($(ProjectDir),$(IntDir)))vcpkg.applocal.log" Encoding="Unicode"/>
        <Exec Condition="$(VcpkgConfiguration.StartsWith('Debug'))"
          Command="$(SystemRoot)\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -noprofile -File %22$(MSBuildThisFileDirectory)applocal.ps1%22 %22$(TargetPath)%22 %22$(VcpkgRoot)debug\bin%22 %22$(TLogLocation)$(ProjectName).write.1u.tlog%22 %22$(IntDir)vcpkg.applocal.log%22"
          StandardOutputImportance="Normal">
        </Exec>
        <Exec Condition="$(VcpkgConfiguration.StartsWith('Release'))"
          Command="$(SystemRoot)\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -noprofile -File %22$(MSBuildThisFileDirectory)applocal.ps1%22 %22$(TargetPath)%22 %22$(VcpkgRoot)bin%22 %22$(TLogLocation)$(ProjectName).write.1u.tlog%22 %22$(IntDir)vcpkg.applocal.log%22"
          StandardOutputImportance="Normal">
        </Exec>
        <ReadLinesFromFile File="$(IntDir)vcpkg.applocal.log">
          <Output TaskParameter="Lines" ItemName="VcpkgAppLocalDLLs" />
        </ReadLinesFromFile>
        <Message Text="@(VcpkgAppLocalDLLs,'%0A')" Importance="Normal" />
        <ItemGroup>
          <ReferenceCopyLocalPaths Include="@(VcpkgAppLocalDLLs)" />
        </ItemGroup>
      </Target>
    </Project>

As you can see, one can easily add a directory in the appropriate section.

**Comment by SeanFarrow** [comment]
Would there be a way to provide a port with the capability to add folders to any of the paths automatically for a specific triplet?
Just thinking out loud.

**Comment by pstieber** [comment]
I fixed this with the attached patch, but it seems really odd to have this for wxWidgets only.

[wxincludepathpatch.txt](https://github.com/Microsoft/vcpkg/files/2362746/wxincludepathpatch.txt)

I changed the additional include directories...

```
    <ClCompile>
      <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)debug\lib\mswud</AdditionalIncludeDirectories>
      <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Release'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)lib\mswu</AdditionalIncludeDirectories>
    </ClCompile>
    <ResourceCompile>
      <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)debug\lib\mswud</AdditionalIncludeDirectories>
      <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Release'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)lib\mswu</AdditionalIncludeDirectories>
    </ResourceCompile>
```

...and it worked.  Does this merit a pull request?

**Comment by degski** [comment]
@pstieber I would say so, because this weirdness has always (as long as I can remember at least) been like that with `wxWidgets` (it seems to be a Windows-related  fudge of some sort, and possibly aims to keep the main include directory clean of Windows-specific stuff).

**Comment by degski** [comment]
@pstieber  Thanks.

**Comment by bsegall** [comment]
> I fixed this with the attached patch, but it seems really odd to have this for wxWidgets only.
> 
> [wxincludepathpatch.txt](https://github.com/Microsoft/vcpkg/files/2362746/wxincludepathpatch.txt)
> 
> I changed the additional include directories...
> 
> ```
>     <ClCompile>
>       <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)debug\lib\mswud</AdditionalIncludeDirectories>
>       <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Release'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)lib\mswu</AdditionalIncludeDirectories>
>     </ClCompile>
>     <ResourceCompile>
>       <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Debug'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)debug\lib\mswud</AdditionalIncludeDirectories>
>       <AdditionalIncludeDirectories Condition="'$(VcpkgNormalizedConfiguration)' == 'Release'">%(AdditionalIncludeDirectories);$(VcpkgRoot)include;$(VcpkgRoot)lib\mswu</AdditionalIncludeDirectories>
>     </ResourceCompile>
> ```
> ...and it worked. Does this merit a pull request?

I have this same problem with wxWidgets, but I'm new to github and vcpkg. Do I just close down VS2017 instances, modify vcpkg.targets as above, then restart VS2017? Or are there more steps involved?

Thanks!

**Comment by pstieber** [comment]
@bsegall You also need to run:

```
.\vcpkg integrate install
```

**Comment by ras0219-msft** [comment]
I think we should fix this issue by moving those headers into the `include\` directory in the port for wxwidgets; would anyone be willing to submit a PR for that?

**Comment by malkia** [comment]
I "fixed" this locally by copying lib/mswu/wx/setup.h into include/wx/setup.h
the difference with debug/lib/mswud/wx/setup.h seems to be only in 

#define wxINSTALL_PREFIX "D:/p/vcpkg/packages/wxwidgets_x64-windows-static-msvcrt/debug"
(release one would be)
#define wxINSTALL_PREFIX "D:/p/vcpkg/packages/wxwidgets_x64-windows-static-msvcrt"

which I'm not sure how reliable would be anyway, as it's always pointing to where vcpkg was compiling things...

(the "windows-static-msvcrt" is my own triplet).

**Comment by ras0219-msft** [comment]
Should be fixed by #4846, thanks everyone for pitching in!
